operator/doublewrite/metricreporter.go

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package doublewrite
+
+import (
+	"context"
+	"time"
+
+	"github.com/cilium/cilium/pkg/allocator"
+	"github.com/cilium/cilium/pkg/identity/key"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/identitybackend"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/hive/cell"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/identity/cache"
+	"github.com/cilium/cilium/pkg/idpool"
+	"github.com/cilium/cilium/pkg/kvstore"
+	kvstoreallocator "github.com/cilium/cilium/pkg/kvstore/allocator"
+	"github.com/cilium/cilium/pkg/kvstore/doublewrite"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+// params contains all the dependencies for the double-write-metric-reporter.
+// They will be provided through dependency injection.
+type params struct {
+	cell.In
+
+	Logger    logrus.FieldLogger
+	Lifecycle cell.Lifecycle
+
+	Clientset k8sClient.Clientset
+
+	Cfg Config
+
+	Metrics *Metrics
+}
+
+type DoubleWriteMetricReporter struct {
+	logger logrus.FieldLogger
+
+	interval time.Duration
+
+	kvStoreBackend            allocator.Backend
+	clientset                 k8sClient.Clientset
+	crdBackend                allocator.Backend
+	crdBackendWatcherStopChan chan struct{}
+
+	mgr *controller.Manager
+
+	metrics *Metrics
+}
+
+func registerDoubleWriteMetricReporter(p params) {
+	doubleWriteMetricReporter := &DoubleWriteMetricReporter{
+		logger:    p.Logger,
+		interval:  p.Cfg.Interval,
+		clientset: p.Clientset,
+		metrics:   p.Metrics,
+	}
+	p.Lifecycle.Append(doubleWriteMetricReporter)
+}
+
+func (g *DoubleWriteMetricReporter) Start(ctx cell.HookContext) error {
+	if option.Config.IdentityAllocationMode != option.IdentityAllocationModeDoubleWriteReadKVstore && option.Config.IdentityAllocationMode != option.IdentityAllocationModeDoubleWriteReadCRD {
+		return nil
+	}
+	g.logger.Info("Starting the Double Write Metric Reporter")
+
+	kvStoreBackend, err := kvstoreallocator.NewKVStoreBackend(kvstoreallocator.KVStoreBackendConfiguration{BasePath: cache.IdentitiesPath, Suffix: "", Typ: nil, Backend: kvstore.Client()})
+	if err != nil {
+		g.logger.WithError(err).Error("Unable to initialize kvstore backend for the Double Write Metric Reporter")
+		return err
+	}
+	g.kvStoreBackend = kvStoreBackend
+
+	crdBackend, err := identitybackend.NewCRDBackend(identitybackend.CRDBackendConfiguration{Store: nil, Client: g.clientset, KeyFunc: (&key.GlobalIdentity{}).PutKeyFromMap})
+	if err != nil {
+		g.logger.WithError(err).Error("Unable to initialize CRD backend for the Double Write Metric Reporter")
+		return err
+	}
+	g.crdBackend = crdBackend
+	// Initialize the CRD backend store
+	g.crdBackendWatcherStopChan = make(chan struct{})
+	go g.crdBackend.ListAndWatch(context.Background(), doublewrite.NoOpHandler{}, g.crdBackendWatcherStopChan)
```

Comment:
I'd suggest adding a waitgroup to ensure that this goroutine terminated when `Stop()` is called. In general, cells should ensure that no goroutines are still running after that the `Stop` method completes, to avoid violating the dependency-based start/stop order.